src/test/java/com/viaversion/fabric/common/AddressParserTest.java

```diff
@@ -0,0 +1,253 @@
+     *     <li>{@link AddressParser#toAddress()} matches the expected address <em>without</em> Via metadata.</li>
+     * </ul>
+     */
+    private static void assertFundamentals(AddressParser viaAddr, String address, Integer port) {
+        assertEquals(viaAddr.serverAddress(), address);
+        assertEquals(viaAddr.port(), port);
+
+        // Tests toAddress to ensure it is returning the port as expected for Minecraft's parser.
+        if (port == null) {
+            assertEquals(viaAddr.toAddress(), address);
+        } else {
+            assertEquals(viaAddr.toAddress(), address + ':' + port);
+        }
+    }
+
+    @CheckReturnValue
+    private static Object[] params(String address, String expected, Integer port, ProtocolVersion... versions) {
+        return new Object[]{address, expected, port, versions};
+    }
+
+    @DataProvider
+    public static Object[][] viaAddrProvider() {
+        // NOTE: Before touching values in here, consider making sure that the parser is functioning as expected.
+        // If the semantic change is expected, carry on. Otherwise, fix the parser, not the test.
+        return new Object[][]{
+                // == Suffixes ==
+                params("[::]._v1_8.viafabric", "[::]", null,
+                        ProtocolVersion.v1_8),
+
+                params("localhost._v1_7_2._v1_16_5.viafabric", "localhost", null,
+                        ProtocolVersion.v1_7_2, ProtocolVersion.v1_16_4),
+
+                params("0.0.0.0._v1_8._v1_9.v_1_21.viafabric", "0.0.0.0", null,
+                        ProtocolVersion.v1_8, ProtocolVersion.v1_9, ProtocolVersion.v1_21),
+
+                params("[::1]._v-2.viafabric", "[::1]", null,
+                        ProtocolVersion.getProtocol(-2)),
+
+                params("127.0.0.1.viafabric", "127.0.0.1", null),
+
+                // Ports
+                params("[::].v_1_8.viafabric:25565", "[::]", 25565,
+                        ProtocolVersion.v1_8),
+
+                params("0.0.0.0.v_-2.viafabric:53", "0.0.0.0", 53,
+                        ProtocolVersion.getProtocol(-2)),
+
+                params("localhost.viafabric:853", "localhost", 853),
+
+                // IPv6
+                params("[fe80::1]._v-2.viafabric:8192", "[fe80::1]", 8192,
+                        ProtocolVersion.getProtocol(-2)),
+
+                params("::.v_-2.viafabric", "::", null,
+                        ProtocolVersion.getProtocol(-2)),
+
+                params("fe80::1._v-2.viafabric", "fe80::1", null,
+                        ProtocolVersion.getProtocol(-2)),
+
+
+                // == Prefixes ==
+                params("viafabric.v1.8;[::]", "[::]", null,
+                        ProtocolVersion.v1_8),
+
+                params("viafabric.v1.7.2+v1.16.5;localhost", "localhost", null,
+                        ProtocolVersion.v1_7_2, ProtocolVersion.v1_16_4),
```

Comment:
Does 1.7.x actually work with ViaFabric? current build intentionally did not accept it.

Reply:
It is out of scope for AddressParser to handle I think; I just picked random versions for test cases and rolled with it to make sure it parsed correctly.